### PR TITLE
[WIP] Migrate to reconciler for builds

### LINKED ISSUE #393: Use as much of knative/pkg as is feasible:
- webhook validation
- controller -> reconciler

### PR DESCRIPTION
**Pull request by ImJasonH**
Fixes #393

## Proposed Changes

  * Use pkg reconciler pattern to reconcile builds
  * Delete `pkg/builder/...` -- only the on-cluster implementation remains, it converts Builds into Pods and creates them -- the `nop` builder impl is no longer necessary
  * Validate build after applying template

## WIP

This PR doesn't currently:
  * take into account build timeouts
  * export `ApplyReplacements` like #415 needs

Looking for preliminary feedback on this approach before adding timeout handling.

**Release Note**

```release-note
Use knative/pkg reconciler pattern for builds
```

### CODE REVIEW

File: pkg/reconciler/build/build.go [hunk 1]
```diff
@@ -44,14 +53,16 @@ type Reconciler struct {
...
+	buildTemplatesLister        listers.BuildTemplateLister
+	clusterBuildTemplatesLister listers.ClusterBuildTemplateLister
 
 	// Sugared logger is easier to use but is not as performant as the
 	// raw logger. In performance critical paths, call logger.Desugar()
 	// and use the returned raw logger instead. In addition to the
 	// performance benefits, raw logger also preserves type-safety at
 	// the expense of slightly greater verbosity.
-	Logger *zap.SugaredLogger
+	logger *zap.SugaredLogger
```
Comment (shashwathi) [inline pkg/reconciler/build/build.go]: You could consider including Base Reconciler inside build reconciler like [this](https://github.com/knative/serving/blob/master/pkg/reconciler/v1alpha1/service/service.go#L44-L51). You get declaration cruft for free

File: pkg/reconciler/build/build.go [hunk 2]
```diff
@@ -107,29 +122,205 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
...
+	}
+	return err
+}
+
+func (c *Reconciler) reconcile(ctx context.Context, b *v1alpha1.Build) error {
+	logger := logging.FromContext(ctx)
+
+	// If the build doesn't define a pod name, it means it's new. We should
+	// apply any template and create its pod.
+	if b.Status.Cluster == nil || b.Status.Cluster.PodName == "" {
```
Comment (shashwathi) [inline pkg/reconciler/build/build.go]: How about this

```
func (c *Reconciler) reconcile(ctx context.Context, b *v1alpha1.Build) error {
pod, err := getPod(b)
if err.IsNotFound(err) {
createPod() // create pod and apply template
} else if err != nil  {
  return err
}
// sync status
```

Comment (shashwathi) [inline pkg/reconciler/build/build.go]: First thing to check would be build status whether its completed or not. If build is completed, ` b.Status.Cluster.PodName` will be set but `GetPod()` will fail with not found error.

File: pkg/reconciler/build/build.go (hunk 1 above)
Comment (mattmoor) [reply pkg/reconciler/build/build.go]: Unfortunately we define that base in `knative/serving` :-/

**Pull request closed without merging**